### PR TITLE
tetragon: Teach file writer to exit gracefully

### PR DESCRIPTION
**Pull request by jrfastab**
We observed that rarely if we SIGTERM Tetragon (e.g. what pod restart and destroy do before SIGKILL it) that a corrupted json file might get left behind. The theory is this is do to writers being killed in the middle of an operation. 

To fix we do two things. First propagate a single context through the processManager and Server objects instead of just using context.Backgroun(). This allows the cancel() call from our SIGTERM handler to be handled by ctx.Done() in the respective go routines. Then we add a waitGroup to the signal handler to ensure it waits until the writer has exited.

I only fix the go routine responsible for writes here. We should audit the other routines as well.

And second I started thinking about how we nest the various objects, at the moment we pass a lot of input types through the New() constructors. But this looked bigger than a fix so held off for now on this.

### CODE REVIEW

File: pkg/server/server.go [hunk 1]
```diff
@@ -90,16 +95,16 @@ func (s *Server) removeNotifierAndDrain(l *getEventsListener) {
 	}
 }
 func (s *Server) GetEvents(request *tetragon.GetEventsRequest, server tetragon.FineGuidanceSensors_GetEventsServer) error {
-	return s.GetEventsWG(request, server, nil)
+	return fmt.Errorf("GetEvents not implemented")
```
Comment (michi-covalent) [inline pkg/server/server.go:98]: why is it not implemented?

Comment (jrfastab) [reply pkg/server/server.go:98]: I don't think its ever used?

Comment (jrfastab) [reply pkg/server/server.go:98]: But maybe to answer the question, I didn't have a correct ctx to add to it so I figured better to just make it bail out.

File: pkg/server/server.go [hunk 2]
```diff
@@ -140,8 +142,8 @@ func (s *Server) GetEventsWG(request *tetragon.GetEventsRequest, server tetragon
 					return err
 				}
 			}
-		case <-server.Context().Done():
-			return server.Context().Err()
+		case <-s.ctx.Done():
+			return s.ctx.Err()
```
Comment (kkourt) [inline pkg/server/server.go:152]: Don't we need to check both contexts here?

I think the server.Context() can also be completed (e.g., due to timeouts in the stream)

File: pkg/server/server.go (hunk 1 above)
Comment (michi-covalent) [reply pkg/server/server.go:98]: hmm i think this is the grpc handler for GetEvents requests, so i'm guessing `getevents` command will break with this change: https://github.com/cilium/tetragon/blob/df691e27c5782041a935f3191a3a6067467d8304/cmd/tetra/getevents/getevents.go#L23

File: pkg/server/server.go [hunk 3]
```diff
@@ -51,9 +53,10 @@ type getEventsListener struct {
 	events chan *tetragon.GetEventsResponse
 }
 
-func NewServer(ctx context.Context, notifier notifier, observer observer) *Server {
+func NewServer(ctx context.Context, wg *sync.WaitGroup, notifier notifier, observer observer) *Server {
 	return &Server{
 		ctx:      ctx,
+		wg:       wg,
```
Comment (kkourt) [inline pkg/server/server.go:59]: How about renaming `wg` to `ctxCleanupWG` or something similar to make it clear that it's the waitgroup for performing any cleanup necessary if `ctx` is completed (assuming my understanding is correct).

File: pkg/server/server.go (hunk 1 above)
Comment (willfindlay) [reply pkg/server/server.go:98]: Yeah this will break grpc :open_mouth:

Comment (jrfastab) [reply pkg/server/server.go:98]: ah OK will fix.

Comment (jrfastab) [reply pkg/server/server.go:98]: Fixed up well enough to not break existing stuff we might need a bit more to pass a closer through this API. But for now I'm going to leave it as is.

File: pkg/server/server.go (hunk 2 above)
Comment (jrfastab) [reply pkg/server/server.go:152]: yep done now.

**Review by kkourt — APPROVED** [review]

**Review by willfindlay — APPROVED** [review]

**Review by michi-covalent — APPROVED** [review]

**Pull request merged**